### PR TITLE
fix(ts): switch from babel preset to ts-loader

### LINKED ISSUE #4361: feat: honor .ts/.tsx pages
### What problem does this feature solve?

This will allow us to have better DX when developing application with Nuxt.js and be able to use these file extensions.

Currently Nuxt pages only pick up .js/.vue files.

### What does the proposed changes look like?

Since Nuxt.js `2.0` is already using Babel 7. I currently manage to use Typescript by just adding `@babel/preset-typescript` and a few other babel plugins.

**But** Nuxt only pick up pages with .js/.vue files.

```vue
<script lang="ts">
// this works by adding those typescript preset & plugins
<script>
```

Nuxt.js needs to allow pages with .ts/.tsx also.

<div align="right"><sub><em>This feature request is available on <a href="https://cmty.app/nuxt">Nuxt</a> community (<a href="https://cmty.app/nuxt/nuxt.js/issues/c8168">#c8168</a>)</em></sub></div>

### PR DESCRIPTION
**Pull request by kevinmarrec**
## Types of changes

- [x] Bug fix (a non-breaking change which fixes an issue)
- [ ] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description
This PR follows #4557.

Instead of creating an issue, I wanted to prove that the `@babel/preset-typescript` is buggy with `.vue` files so I updated a test with a common TypeScript behavior. Check the CI pipeline, something went wrong ! [Shortcut to the failed build](https://circleci.com/gh/nuxt/nuxt.js/19106?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) 😉

A solution will be provided in the next commits and this description will be updated.

*Solution which will be implemented : Replace the `@babel/preset-typescript` setup by a `ts-loader` setup. Please check (https://github.com/vuejs/vue-cli/issues/2371#issuecomment-416522483) and also (https://github.com/TypeStrong/ts-loader#appendtsxsuffixto-regexp-default)*

Resolves #4361

## Checklist:
- [x] Make the tests fail to proove something is wrong 😥
- [x] Implement a solution that make it pass 😎
- [x] Improve the TypeScript test fixture with more TypeScript behaviors 🔨

### CODE REVIEW

File: packages/webpack/src/config/base.js [hunk 1]
```diff
@@ -393,7 +396,7 @@ export default class WebpackBaseConfig {
         hints: this.options.dev ? false : 'warning'
       },
       resolve: {
-        extensions: extensionsToResolve,
+        extensions: ['.wasm', '.mjs', '.js', '.json', '.vue', '.jsx', 'ts'],
```
[the remark this answers is not in the document]
Comment (kevinmarrec) [reply packages/webpack/src/config/base.js]: Good catch ! Yes but how did the tests pass 🤔🤔🤔, i think webpack handle with/without dot. If it doesn't i'm confused 😁

[the remark this answers is not in the document]
Comment (kevinmarrec) [reply packages/webpack/src/config/base.js]: (or it's maybe only for hot reload)

**Review by TheAlexLichter — DISMISSED** [review]

**Review by pi0 — DISMISSED** [review]

**Review by clarkdo — APPROVED** [review]

**Comment by husayt** [comment]
What is the  correct config to switch this on?

**Comment by husayt** [comment]
@kevinmarrec from what I see, it is still required to have a typescript module to use typescript:

```
const ForkTsCheckerWebpackPlugin = require('fork-ts-checker-webpack-plugin')

export default function typeScriptModule() {
  // Add .ts extension for store, middleware and more
  this.nuxt.options.extensions.push("ts")
  // Extend build
  this.extendBuild(config => {
    // Add TypeScript loader
    config.module.rules.push({
      test: /\.ts$/,
      loader: 'ts-loader',
      options: {
        transpileOnly: true,
        appendTsSuffixTo: ['\\.vue$']
      }
    })

    // Add .ts extension in webpack resolve
    if (config.resolve.extensions.indexOf(".ts") === -1) {
      config.resolve.extensions.push(".ts")
    }

    config.plugins.push(new ForkTsCheckerWebpackPlugin({
      vue: true,
      tslint: true,
      formatter: 'codeframe',
      // https://github.com/TypeStrong/ts-loader#happypackmode-boolean-defaultfalse
      checkSyntacticErrors: process.env.NODE_ENV === 'production'
    }))
  })
}
```

**Comment by hartmut-co-uk** [comment]
note @husayt there's still some WIP around ts support
but for nuxt-edge right now, I think the steps roughly are:
- latest nuxt-edge
- `yarn add typescript -D`
- `yarn add ts-loader -D`
- `yarn add fork-ts-checker-webpack-plugin -D`
- `yarn add tslint -D`
- `yarn add @types/node`
- add `./tsconfig.json`
- in `nuxt.config.js` add
```
build: {
  typescript: true
}
```

**Comment by hartmut-co-uk** [comment]
cc https://github.com/nuxt/create-nuxt-app/pull/152

**Comment by hartmut-co-uk** [comment]
apologies - I forgot to add the module - a slimmed down version of what @husayt had provided already - is actually still needed at this stage.
Again to highlight here - I think @kevinmarrec is working on another PR to improve the loader setup + wiring up / integrating `fork-ts-checker-webpack-plugin` + potentially providing full TSX support?
So probably best to wait for him to report back.. 💪 🥇

**Comment by husayt** [comment]
Thank you. Seems like we are almost there. @kevinmarrec shout out  if you
need any support. This is great stuff.

**Comment by kevinmarrec** [comment]
@hartmut-co-uk @husayt 

`fork-ts-checker-webpack-plugin` will be handled by Nuxt with a Nuxt option to toggle it on/off (to not break projects), I'm on it.

And my current TSX support status is really buggy as it need to make files handled by both `ts-loader` for the TypeScript stuff and `babel-loader` for *SX things. I've a setup working but with a Vue warn during tests, so I think there is something wrong.

In the coming hours I'll put the TSX support apart and focus on the TypeScript Checker.

Furthermore, don't worry, there will be a documentation about how to setup your TS project (dependencies and files needed) :)

**Comment by aantipov** [comment]
I just set up to use `@babel/preset-typescript` to transpile my vuejs-typescript code (do not use `nuxt` though). It works fine for me.

But this PR says that `@babel/preset-typescript` is buggy and doesn't work fine. So I started to worry

May I know what exactly `@babel/preset-typescript` doesn't handle correctly?

Is there an issue opened for `@babel/preset-typescript` so that we can track it and know when it's fine to use `@babel/preset-typescript`?

**Comment by kevinmarrec** [comment]
@aantipov See https://github.com/nuxt/nuxt.js/pull/4557#issuecomment-447643605

You're probably using only `.ts` or `.tsx`, or didn't try to use TypeScript in `.vue` SFC files.
TypeScript was (and probably sitll is) not working in Vue SFC.

That's why official Vue CLI TypeScript plugin is also using `ts-loader`.

**Comment by aantipov** [comment]
@kevinmarrec we do use `.vue` files and use 
```
<script lang="ts">...</script>
```
inside of them and it works fine with `@babel/preset-typescript`.
I don't use `vue-property-decorator` and `vue-class-component` either. So maybe it was a problem with just this particular syntax?
I'm just curious, if there is a problem with `@babel/preset-typescript`, then if anyone opened an issue to fix it?
I checked babel issues with `typescript` label and couldn't find anything similar.

**Comment by kevinmarrec** [comment]
@aantipov Indeed maybe the issue is only using classes, but as it's common usage, we can't have it "not working", so we're sticking with `ts-loader`.

Furthermore, we use `ts-loader` in combination to `babel-loader`, so it's still optimized (`ts-loader` only transpile to non-typescript code, then `babel-loader` transpile to the es5).

Also see https://github.com/vuejs/vue-cli/issues/2371#issuecomment-416522483 (old but IMO still true about the unstability regarding Vue + TypeScript when using `@babel/preset-tpyescript` as it doesn't seem to fit all user cases).

**Comment by aantipov** [comment]
@kevinmarrec I got your point.

I just want us to be specific. 

If we say that a tool is buggy and unstable, I'd like to see specific issues proving the point.

Right now I can see there are only [34](https://github.com/babel/babel/labels/area%3A%20typescript) opened babel issues labeled `typescript` and [14](https://github.com/babel/babel/issues?q=is%3Aopen+label%3A%22i%3A+bug%22+label%3A%22area%3A+typescript%22) of them also labeled as `bug`.
Which of them are critical for us so that we can't use `@babel/preset-tpyescript`?

As a side note, Typescript itself contains [1109](https://github.com/Microsoft/TypeScript/issues?q=is%3Aopen+is%3Aissue+label%3ABug)(!!!) opened issues labeled as `bug` and it doesn't stop us using it.

As of a case with  `vue-property-decorator` and `vue-class-component`, I will try to check how `@babel/preset-tpyescript` works here

**Comment by kevinmarrec** [comment]
@kevinmarrec The problem is not `babel` issue regarding `typescript`, it's specific to `.vue` files, which is kind of out of `babel` scope. The `@babel/preset-typescript` seems just not able to transpile TypeScript classes in `.vue` files.

Nuxt chose to go in the same way than Vue CLI TypeScript plugin which means using `ts-loader` + `fork-ts-checker-webpack-plugin`.

If official Vue CLI TypeScript plugin switch to `@babel/preset-typescript`, we'll probably do as well, but for now it's **not stable for Vue**. It's stable for almost all use cases, but the case with classes tags it as not stable.

You might as well noticed that `@babel/preset-typescript` uses https://babeljs.io/docs/en/babel-plugin-transform-typescript which has few known caveats.

**Comment by aantipov** [comment]
@kevinmarrec Thanks a lot for answering my questions and concerns and for clarifying them.

>If official Vue CLI TypeScript plugin switch to @babel/preset-typescript, we'll probably do as well, but for now it's not stable for Vue.

👍 got it

>You might as well noticed that @babel/preset-typescript uses https://babeljs.io/docs/en/babel-plugin-transform-typescript which has few known caveats.

Yes, those caveats should be OK for most users as far as they are informed about them.
TSLint has a predefined `no-namespace` rule, so the use of them is against best practices.
`import =` and `export =` is a deprecated syntax.

>It's stable for almost all use cases, but the case with classes tags it as not stable.

Ok, will try to dive deeper what is happening with classes and what's the problem with `babel`+`.vue` (in theory, `vue-loader` should pass `babel` only `<script>` part of the file, so it should work the same way as with `.ts` files)

**Comment by kevinmarrec** [comment]
Thanks @aantipov for your time, let us know if you figure out what's going on regarding the issue using `@babel/preset-typescript`.

**Pull request merged**